File: 0418-sentence-screen-fitting/0418-sentence-screen-fitting.js
/**
 * @param {string[]} sentence
 * @param {number} rows
 * @param {number} cols
 * @return {number}
 */
var wordsTyping = function(sentence, rows, cols) {
    
    const n = sentence.length;
    const cache = new Array(n);
    
    for(let i = 0; i < n; ++i) {
        let index = i;
        let len = 0;
        let wordNum = 0;
        while((len + sentence[index % n].length) <= cols) {
            len += sentence[index % n].length + 1;
            wordNum++;
            index++;
        }
        cache[i] = wordNum;
    }
    
    
    let words = 0; let index = 0;
    for(let i = 0; i < rows; ++i) {
        words += cache[index];
        index = (cache[index] + index) % n;
    }
    
    return Math.floor(words / n);
};

File: 0418-sentence-screen-fitting/0418-sentence-screen-fitting.py
class Solution:
    def wordsTyping(self, sentence: List[str], rows: int, cols: int) -> int:
        n = len(sentence)
        cache = [0] * n
        
        for i in range(n):
            index, wordNum, length = i, 0, 0
            while length + len(sentence[index % n]) <= cols:
                length += len(sentence[index % n]) + 1
                index += 1
                wordNum += 1
            cache[i] = wordNum
            
        words, index = 0, 0
        for i in range(rows):
            words += cache[index]
            index = (index + cache[index]) % n
        
        return words // n
        